src/coreclr/jit/rationalize.cpp

```diff
@@ -565,6 +625,12 @@ Compiler::fgWalkResult Rationalizer::RewriteNode(GenTree** useEdge, Compiler::Ge
             RewriteAssignment(use);
             break;
 
+#ifdef TARGET_ARM64
+        case GT_SUB:
+            RewriteSubLshDiv(use);
+            break;
+#endif
+
```

Comment:
If it needs to be done in preorder, then there needs to be additional logic to handle the interference checks, right? Because putting it in pre-order won't match the desired shape in scenarios that have COMMAS.

Reply:
If it's done in post-order then it needs interference checks, precisely because you won't see the commas as operands -- those commas contain embedded statements that are linearized and that you need to check interference with in post-order.

In what situations are you seeing commas where the transform would be able to kick in? That sounds a little surprising to me.

Reply:
[asmdiffs.zip](https://github.com/dotnet/runtime/files/8721040/asmdiffs.zip)

Reply:
FWIW, it's not hard to do the interference check that I expect will catch all cases we care about -- you can just check if the range of the `sub` is closed.

Reply:
Example program:
```csharp
using System;
using System.Runtime.CompilerServices;

namespace Benchstone.BenchI
{
    public static class TreeSort
    {

#if DEBUG
    public const int Iterations = 1;
#else
        public const int Iterations = 1200;
#endif

        const int SortElements = 5000;
        const int Modulus = 65536;

        sealed class Node
        {
            public Node Left;
            public Node Right;
            public int Val;
            public Node(int n)
            {
                Left = null;
                Right = null;
                Val = n;
            }
        }

        static int s_biggest;
        static int s_littlest;
        static int s_seed;

        static void InitRand()
        {
            s_seed = 33;
        }

        static int Rand(ref int seed)
        {
            int multiplier = 25173;
            int increment = 13849;
            seed = (multiplier * seed + increment) % Modulus;
            return seed;
        }

        static void InitArray(int[] sortList)
        {
            InitRand();
            s_biggest = 0;
            s_littlest = Modulus;
            for (int i = 1; i <= SortElements; i++)
            {
                sortList[i] = Rand(ref s_seed) - 1;
                if (sortList[i] > s_biggest)
                {
                    s_biggest = sortList[i];
                }
                else if (sortList[i] < s_littlest)
                {
                    s_littlest = sortList[i];
                }
            }
        }

        static void Insert(int n, Node t)
        {
            if (n > t.Val)
            {
                if (t.Left == null)
                {
                    t.Left = new Node(n);
                }
                else
                {
                    Insert(n, t.Left);
                }
            }
            else if (n < t.Val)
            {
                if (t.Right == null)
                {
                    t.Right = new Node(n);
                }
                else
                {
                    Insert(n, t.Right);
                }
            }
        }

        static bool CheckTree(Node p)
        {
            bool result = true;
            if (p.Left != null)
            {
                if (p.Left.Val <= p.Val)
                {
                    result = false;
                }
                else
                {
                    result &= CheckTree(p.Left);
                }
            }

            if (p.Right != null)
            {
                if (p.Right.Val >= p.Val)
                {
                    result = false;
                }
                else
                {
                    result &= CheckTree(p.Right);
                }
            }

            return result;
        }

        static bool Trees(int[] sortList)
        {
            InitArray(sortList);
            Node tree = new Node(sortList[1]);
            for (int i = 2; i <= SortElements; i++)
            {
                Insert(sortList[i], tree);
            }
            bool result = CheckTree(tree);
            return result;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static bool Bench()
        {
            int[] sortList = new int[SortElements + 1];
            bool result = Trees(sortList);
            return result;
        }

        static bool TestBase()
        {
            bool result = true;
            for (int i = 0; i < Iterations; i++)
            {
                result &= Bench();
            }
            return result;
        }

        public static int Main()
        {
            bool result = TestBase();
            return (result ? 100 : -1);
        }
    }
}

```

Looking at part of `InitArray` :
```diff
 G_M14667_IG04:        ; gcrefRegs=0000 {}, byrefRegs=0028 {x3 x5}, byref
             str     w0, [x1,#-16]
             b       G_M14667_IG06
@@ -104,12 +104,15 @@ G_M14667_IG08:        ; gcrefRegs=0001 {x0}, byrefRegs=0000 {}, byref, isz
             mov     w5, #0xd1ffab1e
             mov     w6, #0xd1ffab1e
             madd    w4, w4, w5, w6
-            and     w5, w4, #0xd1ffab1e
-            negs    w4, w4
-            and     w4, w4, #0xd1ffab1e
-            csneg   w4, w5, w4, mi
+            asr     w5, w4, #31
+            and     w5, w5, #0xd1ffab1e
+            add     w5, w5, w4
+            asr     w5, w5, #16
+            lsl     w5, w5, #16
+            sub     w4, w4, w5
             str     w4, [x3]
-            sub     w3, w4, #1
+            ldr     w3, [x3]
+            sub     w3, w3, #1
             mov     w4, w3
             ldr     w5, [x0,#8]
             cmp     w2, w5
@@ -121,7 +124,7 @@ G_M14667_IG08:        ; gcrefRegs=0001 {x0}, byrefRegs=0000 {}, byref, isz
             ; byrRegs -[x5]
             cmp     w3, w5
             ble     G_M14667_IG10
-						;; size=88 bbWeight=0.04 PerfScore 0.88
+						;; size=100 bbWeight=0.04 PerfScore 1.10
```

Reply:
BTW, you can do this in `fgSimpleLowering` instead without having to move that phase. That makes the change a little less controversial :)

Reply:
Doing it in simple lowering makes sense, though the optimization isn't exactly "simple" :)